Static Class and Helper Functions

stephen-hill opened the pull request:
This pull request adds the following classes:
- StephenHill\StaticBase58
- \Base58

and adds the following functions:
- base58_encode
- base58_decode

This pull request is directly related to issue #14.

The master branch will be tagged with v1.2 when this branch is merged.

Still todo is document the changes in readme.md.

stephen-hill committed:
New class for calling encode and decode statically.

stephen-hill committed:
New helper file to create the namespaceless Base58 class, and to create the base58_encode and base58_decode functions.

stephen-hill committed:
Tests for the new static classes and functions.

barbushin commented on base58.php:
```diff
@@ -0,0 +1,33 @@
+<?php
+
+if (false === function_exists('base58_encode'))
```
No need in `false ===`, it looks weird :D

barbushin commented on base58.php:
```diff
@@ -0,0 +1,33 @@
+<?php
```
I guess you can add `use StephenHill\StaticBase58` to access class name `StaticBase58` without namespace.

barbushin commented on base58.php:
```diff
@@ -0,0 +1,33 @@
+    }
+}
+
+if (false === class_exists('Base58', true))
```
Same. Just use

    if (!class_exists('Base58', true))

barbushin commented on src/StaticBase58.php:
```diff
@@ -0,0 +1,28 @@
+
+namespace StephenHill;
+
+class StaticBase58
```
If this is singleton, then you should add `private function __construct() {}`.

barbushin commented on src/StaticBase58.php:
```diff
@@ -0,0 +1,28 @@
+        return self::getInstance()->decode($string);
+    }
+
+    protected static function getInstance()
```
I think it's better to make it public, so users can still configure this global instance. Actually, it's very strange that `Base58` class does not have methods like `setAlphabet()` and `setMathService()`. I think it's not very good practice to implement everything in `__construct()`.

barbushin commented on src/StaticBase58.php:
```diff
@@ -0,0 +1,28 @@
+
+    protected static function getInstance()
+    {
+        if (null === static::$instance)
```
`if (!static::$instance)` is enough

barbushin commented on src/StaticBase58.php:
```diff
@@ -0,0 +1,28 @@
+
+    public static function encode($string)
+    {
+        return self::getInstance()->encode($string);
```
`return static::...` is better if somebody decide to override `getInstance()` to use another `Base58` implementation class`

barbushin commented on src/StaticBase58.php:
```diff
@@ -0,0 +1,28 @@
+
+    public static function decode($string)
+    {
+        return self::getInstance()->decode($string);
```
`return static::...` is better if somebody decide to override `getInstance()` to use another `Base58` implementation class`

barbushin commented:
Also I think there should be some usage example in README.md.

In common, looks great! :+1:

stephen-hill committed:
Added `use` statement because the class is referenced multiple times.

stephen-hill committed:
Added the ability to get and replace the current service provider. Need to
add unit tests for these methods.

stephen-hill committed:
Made getInstance() public. Added a public setInstance() method. Need to
document methods and add unit tests.

stephen-hill committed:
Generate code coverage from root. Exclude tests directory.

stephen-hill committed:
Include the php file that defines the global base58 functions and static classes.

stephen-hill committed:
Exclude generated code coverage reports.

stephen-hill committed:
New test for the getService() and replaceService() methods.

jasny commented:
You shouldn't really define these functions in the global namespace. Put them in the `StephenHill` namespace instead and put the following on top of your source file.

```php
use function StephenHill\base58_encode;
use function StephenHill\base58_decode;
```
